[JSC] Implement `Array.prototype.fill` in C++

sosukesuzuki opened the pull request:
#### 619232f3b3be51ce68b813947cdaf3e414a4d3e6
<pre>
[JSC] Implement `Array.prototype.fill` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=283255">https://bugs.webkit.org/show_bug.cgi?id=283255</a>

Reviewed by Yusuke Suzuki and Keith Miller.

The current implementation of Array#fill is written in simple JavaScript. This patch implements it
in C++.

                                         TipOfTree                  Patched

array-prototype-fill-undecided         5.3186+-0.0609     ^      1.1502+-0.0220        ^ definitely 4.6240x faster
array-prototype-fill-int32             4.0577+-0.4950     ^      1.1149+-0.0158        ^ definitely 3.6395x faster
array-prototype-fill-contiguous        5.6943+-0.3723     ^      1.1534+-0.0211        ^ definitely 4.9368x faster
array-prototype-fill-double            3.6932+-0.1337     ^      1.1045+-0.0155        ^ definitely 3.3438x faster

* JSTests/microbenchmarks/array-prototype-fill-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-undecided.js: Added.
(test):
* JSTests/stress/array-prototype-fill-fast.js: Added.
(sameArray):
(throw.new.Error):
* JSTests/stress/array-prototype-fill-side-effects.js: Added.
(shouldBe):
(set value):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(fill):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fillMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: <a href="https://commits.webkit.org/287215@main">https://commits.webkit.org/287215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf953f1571d0b7ff1580a741d5c03d1bc5e64a1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78601 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57646 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83262 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29866 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66797 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61567 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19490 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81668 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51581 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/70103 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41877 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48927 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25465 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28204 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71747 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70029 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84630 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77838 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4101 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69793 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6127 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67561 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69047 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17223 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13071 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11556 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100146 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5913 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21874 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5900 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9335 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/ArrayPrototype.js:
```diff
@@ -231,6 +231,8 @@ function fill(value /* [, start [, end]] */)
+        return array;
     for (; k < final; k++)
         array[k] = value;
     return array;
```
I feel like we should implement `Array#fill` totally in C++.

sosukesuzuki replied:
fixed

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,69 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+
+    auto thisValue = callFrame->thisValue().toThis(globalObject, ECMAMode::strict());
+    RETURN_IF_EXCEPTION(scope, { });
+    if (thisValue.isUndefinedOrNull())
```
Put `UNLIKELY()`

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,69 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    uint64_t length = toLength(globalObject, thisObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    int64_t relativeStart = callFrame->argument(1).toIntegerOrInfinity(globalObject);
```
`toIntegerOrInfinity`'s return value is `double` since it can be infinity. So we cannot simply cast it to int64_t.
Consider using `argumentClampedIndexFromStartOrEnd` as it can handle these things efficiently.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,69 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    if (!end.isUndefined()) {
+        relativeEnd = end.toIntegerOrInfinity(globalObject);
+        RETURN_IF_EXCEPTION(scope, { });
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,69 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    }
+
+    for (; k < final; k++) {
+        thisObject->putByIndexInline(globalObject, static_cast<uint64_t>(k), callFrame->argument(0), false);
```
Instead of calling `callFrame->argument(0)` repeatedly, we should get it once and use a variable.
Also, why is it `false`?

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,69 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+
+    if (isJSArray(thisValue)) {
+        auto* array = jsCast<JSArray*>(thisValue);
+        if (array->fastFill(vm, k, final, callFrame->argument(0)))
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,47 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue argStart = callFrame->argument(1);
+    size_t k = argumentClampedIndexFromStartOrEnd(globalObject, argStart, length, 0);
```
Use `uint64_t`.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,47 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue argEnd = callFrame->argument(2);
+    size_t final = argumentClampedIndexFromStartOrEnd(globalObject, argEnd, length, length);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,65 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+            return NonArray;
+        }
+    };
+    IndexingType nextType = getNextIndexingType();
```
Let's do it in,
```
IndexingType nextType = ([type, value]() {
   ...
})();
```

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,65 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+
+    if (nextType == ArrayWithDouble) {
+        auto* data = butterfly()->contiguousDouble().data();
+        std::fill(data + startIndex, data + endIndex, value.asNumber());
```
This is fine, but you can use memset_pattern8 instead.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,65 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+        std::fill(data + startIndex, data + endIndex, value.asNumber());
+    } else {
+        auto* data = butterfly()->contiguous().data();
+        std::fill(data + startIndex, data + endIndex, std::bit_cast<const WriteBarrier<Unknown>>(JSValue::encode(value)));
```
For `ArrayWithInt32` case, this is fine, but for ArrayWithContiguous case, this is not correct since concurrent GC can scan this butterfly while writing, and if the written value is not fully commited (like, byte-by-byte write), then it becomes broken cell in the concurrent GC marker.
You need to do something like `gcSafeMemcpy` but for memset.

sosukesuzuki replied:
For MacOS, can we use `memset_pattern8` here?

sosukesuzuki replied:
Implementing a GC safe memset is difficult for me, so I used `memset_pattern8` on MacOS and a for loop for other OS.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,74 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+#if OS(DARWIN)
+        auto* data = butterfly()->contiguous().data();
+        auto pattern = std::bit_cast<const WriteBarrier<Unknown>>(JSValue::encode(value));
+        memset_pattern8(data + startIndex, &pattern, sizeof(JSValue) * (endIndex - startIndex));
```
No, memset_pattern8 does not have that guarantee. So for now, use loop. And add FIXME URL for the future fix.

sosukesuzuki replied:
Thank you, I've fixed

kmiller68 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1900,6 +1901,47 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncConcat, (JSGlobalObject* globalObject, Ca
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue argEnd = callFrame->argument(2);
+    uint64_t final = argumentClampedIndexFromStartOrEnd(globalObject, argEnd, length, length);
```
Huh, I assumed final was a keyword.

sosukesuzuki replied:
Renamed to `finalIndex`

kmiller68 commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,69 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+    } else {
+        // FIXME: https://bugs.webkit.org/show_bug.cgi?id=283786
+        for (unsigned i = startIndex; i < endIndex; ++i)
+            butterfly()->contiguous().at(this, i) = std::bit_cast<const WriteBarrier<Unknown>>(JSValue::encode(value));
```
I would do this as:
```
auto contiguousStorage = butterfly()->contiguous();
for (unsigned i = startIndex; i < endIndex; ++i)
    contiguousStorage.at(this, i).setWithoutWriteBarrier(value);
```

sosukesuzuki replied:
fixed

kmiller68 commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,69 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+
+    IndexingType type = indexingType();
+    IndexingType nextType = [type, value]() {
+        if (!(type & IsArray))
+            return NonArray;
+        if (hasAnyArrayStorage(type))
+            return NonArray;
+        switch (type) {
+        case ArrayWithInt32:
+        case ArrayWithUndecided:
+            if (value.isInt32())
+                return ArrayWithInt32;
+            if (value.isNumber())
+                return ArrayWithDouble;
+            return ArrayWithContiguous;
+        case ArrayWithDouble:
+            if (value.isNumber())
+                return type;
+            return ArrayWithContiguous;
+        case ArrayWithContiguous:
+            return type;
+        default:
+            return NonArray;
+        }
+    }();
+    if (type == ArrayWithUndecided) {
+        if (nextType == ArrayWithInt32)
+            convertUndecidedToInt32(vm);
+        else if (nextType == ArrayWithDouble)
+            convertUndecidedToDouble(vm);
+        else if (nextType == ArrayWithContiguous)
+            convertUndecidedToContiguous(vm);
+        else {
+            ASSERT_NOT_REACHED();
+            return false;
+        }
+    } else if (type != nextType)
+        return false;
+
+    ASSERT(nextType == indexingType());
```
I think this should be: 
```
IndexingType nextType = leastUpperBoundOfIndexingTypeAndValue(indexingType(), value);
convertToIndexingTypeIfNeeded(nextType);
```
where `JSObject::convertToIndexingTypeIfNeeded` is a big switch on both the old and new indexing types and calling the appropriate `JSObject::convert<type>To<nextType>` function.

sosukesuzuki replied:
Adding such a function is a bit challenging for me at the moment.So, I would like to add it in a separate Pull Request. I have filed a bug report for this( https://bugs.webkit.org/show_bug.cgi?id=283816 ). Would that be acceptable?

kmiller68 replied:
Seems ok.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -481,6 +481,69 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
+#else
+        std::fill(data + startIndex, data + endIndex, value.asNumber());
+#endif
+    } else {
```
We should have `ArrayWithInt32` fast path too. (which can use memset_pattern8).

sosukesuzuki replied:
Added the fast path

Constellation reviewed (approved):
r=me with comments. And let's resolve Keith's comments too.

kmiller68 reviewed (approved):
r=me too.

sosukesuzuki committed:
[JSC] Implement `Array.prototype.fill` in C++
https://bugs.webkit.org/show_bug.cgi?id=283255

Reviewed by Yusuke Suzuki and Keith Miller.

The current implementation of Array#fill is written in simple JavaScript. This patch implements it
in C++.

                                         TipOfTree                  Patched

array-prototype-fill-undecided         5.3186+-0.0609     ^      1.1502+-0.0220        ^ definitely 4.6240x faster
array-prototype-fill-int32             4.0577+-0.4950     ^      1.1149+-0.0158        ^ definitely 3.6395x faster
array-prototype-fill-contiguous        5.6943+-0.3723     ^      1.1534+-0.0211        ^ definitely 4.9368x faster
array-prototype-fill-double            3.6932+-0.1337     ^      1.1045+-0.0155        ^ definitely 3.3438x faster

* JSTests/microbenchmarks/array-prototype-fill-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-fill-undecided.js: Added.
(test):
* JSTests/stress/array-prototype-fill-fast.js: Added.
(sameArray):
(throw.new.Error):
* JSTests/stress/array-prototype-fill-side-effects.js: Added.
(shouldBe):
(set value):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(fill):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fillMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: https://commits.webkit.org/287215@main
